Add TED (The Energy Detective) integration

rianadon opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
Adds a new integration TED that supports both the TED5000 and TED6000 energy monitors.
These are both popular energy monitoring solutions, so the integration will help users looking to integrate these energy monitors into Home Assistant's energy panel.

This integration differs from #54555 in that the old TED5000 integration is kept so that there are no breaking changes with the pull request.

Users using the old TED5000 integration will see the following in their repairs panel:
<img width="643" alt="Screenshot 2022-11-29 at 5 32 34 PM" src="https://user-images.githubusercontent.com/18223213/204686815-116dcf30-40b2-4892-8160-279028384bfe.png">

The features are as follows:
- Shows MTU breakdown of power and voltage (same as the ted5000 integration)
- **NEW**: Shows today's and month-to-date totals of MTU usage (which can be integrated with energy panel)
- **NEW**: Shows system-wide totals of power usage, power returned to grid, and power consumed from grid
- **NEW**: (For TED6000) Shows Spyder breakdown of current, today, and month-to-date total power usage
- **NEW**: Configuration options to create entities for MTU breakdowns and Spyder breakdowns

<img width="868" alt="Screenshot 2022-12-02 at 1 17 11 AM" src="https://user-images.githubusercontent.com/18223213/205259520-5bc2fa8f-5cb6-4dd0-9bc5-0fc3538ce085.png">


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

Brands pull request: https://github.com/home-assistant/brands/pull/3897
Documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19234

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rianadon committed:
Adapt ted5000 to ted6000

rianadon committed:
Add config flow and coordinator similar to enphase integration

rianadon committed:
Clean up config flow

rianadon committed:
Add test

rianadon committed:
Configure with CONF_HOST, remove envoy references

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

rianadon committed:
unit_of_measurement -> native_unit_of_measurement

rianadon committed:
Merge ted5000 & ted6000 + host communication logic on pypi

rianadon committed:
Mock tedpy in tests and upgrade

rianadon committed:
Increase code coverage

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

rianadon committed:
Increase test coverage even more

rianadon committed:
Remove reauth code

rianadon committed:
Remove placeholders since it looks like there's no paths for them

rianadon committed:
Add MTU & Spyder data reporting

rianadon committed:
Add translations

rianadon committed:
Decrease async timeouts

rianadon committed:
Don't publish daily consumption for spyders, state class total

rianadon committed:
Use model instead of ID to name integration

rianadon committed:
Use native_value

rianadon committed:
Allow spyder devices to be configured via options

rianadon committed:
Add test for options flow

rianadon committed:
Upgrade tedpy

rianadon committed:
Remove non-English translations

I think I read somewhere these are automatically generated by the
translation tool so should not be comitted for now?

rianadon committed:
Upgrade tedpy and support mtu historical data

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

rianadon committed:
Remove sensor.py from ted5000

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted6000

Ramias1 committed:
Update sensor.py; minor changes

Changed line 69 to use async_get to suppress log error.  

Changed line 38 state_class to TOTAL_INCREASING so energy monitor will work with energy dashboard.

rianadon committed:
Add ted5000 integration back

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted-new

rianadon committed:
Fix requirements files

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted-new

rianadon committed:
Add issue to ted5000 integration

natethelen commented:
I haven't been able to test it in this branch, but when I copied this ted folder to my dev branch and start the wizard, I get an error that there are conflicting dependencies in tedpy

rianadon commented:
Indeed. It looks like the version of `httpx` in `tedpy` needs to be updated.

rianadon committed:
Upgrade tedpy dependency

rianadon commented on homeassistant/components/ted5000/sensor.py:
```diff
@@ -55,6 +57,20 @@ def setup_platform(
+        hass,
+        DOMAIN,
+        "deprecated_ted5000_integration",
+        breaks_in_ha_version="2023.4.0",
```
This was an arbitrary release. I'm not sure when is the proper time to schedule the removal of `ted5000`.

Quentame replied:
As I remember it's minimum 3 releases after the code release.

So it's good now if this PR is merged for `2023.1.0`

rianadon replied:
Perfect! I'll keep that in mind if the PR is merged after the January release.

rianadon committed:
Fix pylint warnings

natethelen commented:
BTW, I have a fairly large setup with 4 MTUs and 2 spiders on 2, 200 amp main panels, solar, and a 100 amp sub panel. I am running into some problems and will debug in the morning.

realumhelp commented:
> BTW, I have a fairly large setup with 4 MTUs and 2 spiders on 2, 200 amp main panels, solar, and a 100 amp sub panel. I am running into some problems and will debug in the morning.

Some of the Solar stuff we kinda honestly winged since we didn't have the equipment. Did you by chance test the underlying software library standalone to make sure it is pulling the correct data?

https://github.com/rianadon/the-energy-detective-py

I haven't toyed with it for a while but it might be a bit of fun to get back into it!

natethelen commented:
> Some of the Solar stuff we kinda honestly winged since we didn't have the equipment. Did you by chance test the underlying software library standalone to make sure it is pulling the correct data?

I have used tedpy and everything looks good when using it directly.  

The error I am getting is in TedSensor.native_value for the sensor with description of "net_now".  In this case self.coordinator.data.get(key) returns None and an exception happens when trying to getattr of None.  Still digging, but if that sparks something in your brain...

natethelen commented:
> The error I am getting is in TedSensor.native_value for the sensor with description of "net_now". In this case self.coordinator.data.get(key) returns None and an exception happens when trying to getattr of None

It seems that the coordinator has the "net" values named "energy".  I will do a quick change to see if that solves the problem.

natethelen commented:
> It seems that the coordinator has the "net" values named "energy". I will do a quick change to see if that solves the problem.

Changing line 45 of ted/__init__.py to this solves my error:

data["net"] = ted_reader.energy()

Now just waiting to see if the energy graphs look correct...

rianadon committed:
Correctly name data in coordinator

rianadon commented:
Thank you! I took a look again and this seems like the correct fix. I was deciding between giving the net energy sensors ids with `net` or `energy` and chose both I guess. Now they're all named `net`.

rianadon commented:
I also noticed this PR: https://github.com/home-assistant/core/pull/57933. It looks like it adds a few other sensors we don't have in this PR like the number of days left in the billing cycle.

Perhaps it would be good to add those sensors once this PR is merged.

natethelen commented:
I few other things I noticed:
- When there is solar as part of the installation the "net" energy (both the totals and the MTUs) can be negative.  Consequently, the state_class for the daily and mtd sensors cannot be TOTAL_INCREASING and must be TOTAL.
- In the _now values in the SensorEntityDescription you could use device_class=SensorDeviceClass.POWER
- POWER_WATT is deprecated (replaced with UnitOfPower.WATT)
- ENERGY_WATT_HOUR is deprecated (replaced with UnitOfEnergy.WATT_HOUR)

Will continue to test and review...

natethelen commented:
Naming Note:  In a system that has Solar, the .consumption() values are how much total energy used in the house/office/etc, not necessarily the amount "consumed" from the utility.  In this case, .energy(). is the amount consumed from (or returned to, if negative) the utility.  It may be less confusing to use stype of "Energy Usage" for "consumption" and "Energy Consumption" for "net".  Thoughts?

natethelen commented:
I am seeing a good number of these: "Timeout fetching TED TED 6000 data", then recovering the next try.  Maybe with a larger deployment like mine it regularly takes close to 10 seconds.  Bump timeout up to 15 seconds?

rianadon committed:
Remove deprecated units, net = total, rename sensors

rianadon commented:
Thank you for all the detailed feedback and suggestions!

I agree with using "Energy Usage" for "consumption", but I find "Energy Consumption" still too confusing for the amount consumed from the grid. I think either "Grid Energy Consumption" or "Net Grid Energy" (what I'm now using in the code) would be more clear.

natethelen commented:
Looks great.  Ship it! :)

rianadon committed:
Increase timeout

natethelen commented:
FYI, I have been running this pretty consistently for a few days without issue

Quentame commented:
@natethelen : Anyone can review the code of an HA PR using comment/review actions on the "Files changed" tab 😉, not only members of [home-assistant/developers](https://github.com/orgs/home-assistant/teams/developers/members) !

We actualy encourage people to do so ☺️ !

And using those tools (code comment/review) improve following history of the PR.

Quentame commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,150 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Manage the TED options."""
+        return await self.async_step_options()
```
Usually the flow is on the init step, please move it here

Quentame commented on homeassistant/components/ted/__init__.py:
```diff
@@ -0,0 +1,100 @@
+
+            return data
+
+    coordinator = DataUpdateCoordinator(
```
Eventually create a dedicated `TedUpdateCoordinator`:
- https://github.com/home-assistant/core/blob/dev/homeassistant/components/ipp/coordinator.py
- https://github.com/home-assistant/core/blob/dev/homeassistant/components/synology_dsm/coordinator.py

Quentame reviewed (changes requested):
Haven't looked at tests yet

rianadon committed:
Apply suggestions from code review

Co-authored-by: Quentame <polletquentin74@me.com>

rianadon committed:
Move coordinator to its own file

rianadon committed:
Remove show_ from config options

rianadon committed:
Rename sensor description functions

rianadon committed:
Remove name and icon methods from sensor

rianadon committed:
Remove unique_id method

rianadon committed:
Fix imports

SupremeSports commented:
Could you integrate the utility sensors such as actual kWh cost, Time Of Use, etc. I already did it in my now closed PR#82959

`            current_rate = int(doc["LiveData"]["Utility"]["CurrentRate"])
            days_left = int(doc["LiveData"]["Utility"]["DaysLeftInBillingCycle"])
            plan_type = int(doc["LiveData"]["Utility"]["PlanType"])
            plan_type_str = {0: "Flat", 1: "Tier", 2: "TOU", 3: "Tier+TOU"}
            carbon_rate = int(doc["LiveData"]["Utility"]["CarbonRate"])
            read_date = int(doc["LiveData"]["Utility"]["MeterReadDate"])
            gateway_year = int(doc["LiveData"]["GatewayTime"]["Year"]) + 2000
            gateway_month = int(doc["LiveData"]["GatewayTime"]["Month"])
            gateway_day = int(doc["LiveData"]["GatewayTime"]["Day"])
            gateway_hour = int(doc["LiveData"]["GatewayTime"]["Hour"])
            gateway_minute = int(doc["LiveData"]["GatewayTime"]["Minute"])
            gateway_second = int(doc["LiveData"]["GatewayTime"]["Second"])

            gateway_datetime = datetime(gateway_year, gateway_month, gateway_day, gateway_hour, gateway_minute, gateway_second)
            gateway_utc_time = gateway_datetime.replace(tzinfo=get_time_zone(time_zone))

            if plan_type in (0, 2):
                current_tier = 0
            else:
                current_tier = int(doc["LiveData"]["Utility"]["CurrentTier"]) + 1

            if plan_type < 2:
                current_tou = 0
                current_tou_str = "Not Configured"
            else:
                current_tou = int(doc["LiveData"]["Utility"]["CurrentTOU"]) + 1
                current_tou_str = doc["LiveData"]["Utility"]["CurrentTOUDescription"]
`

rianadon committed:
Add test for config options

rianadon commented:
@Quentame Thank you so much for the detailed PR review! I really appreciate all the feedback. I've addressed your comments in the recent commits.

@SupremeSports I have my hands a bit full this week. I'm thinking these sensors would be great to add in a second PR?

SupremeSports commented:
> As per my experience (with my PR) it can be long. Mine took over 15 months and was still not processed. I'd suggest to put everything there all at once...

rianadon committed:
Add spyders and mtus to test

Quentame commented on homeassistant/components/ted/coordinator.py:
```diff
@@ -0,0 +1,68 @@
+
+
+class TedUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching IPP data from single endpoint."""
+
+    def __init__(self, hass: HomeAssistant, *, name: str, ted_reader: TED) -> None:
+        """Initialize global IPP data updater."""
```
Apparently you took the exemple of IPP 😄

rianadon replied:
I made it a bit too obvious, didn't I? Oops.

rianadon committed:
Fix incorrect docstrings

Quentame commented on homeassistant/components/ted/translations/en.json:
```diff
@@ -0,0 +1,36 @@
+                    "mtu_energy_daily": "Create MTU Daily Energy entities",
+                    "mtu_energy_mtd": "Create MTU Month-to-Date Energy entities"
+                },
+                "description": "Configure TED integration",
+                "title": "TED options"
```
This file is not in sync with `strings.json` file.
It also should not be modified manually, but only by a command line (don't remember which).

That said, I don't think options needs a description & title.

rianadon replied:
Found it! `python3 -m script.translations develop`

Quentame replied:
Thanks 😄

rianadon committed:
Fix step in translations, remove title + desc

rianadon committed:
Remove YAML import config flow

Quentame commented on .coveragerc:
```diff
@@ -1287,6 +1287,8 @@ omit =
     homeassistant/components/tautulli/sensor.py
     homeassistant/components/ted5000/sensor.py
+    homeassistant/components/ted/__init__.py
+    homeassistant/components/ted/sensor.py
```
plus coordinator

rianadon committed:
Omit coordinator in coverage

davet2001 commented:
> As per my experience (with my PR) it can be long. Mine took over 15 months and was still not processed. I'd suggest to put everything there all at once...

My suggestion would be to make a PR as small as possible. The short ones are closed quickly.

davet2001 commented on homeassistant/components/ted/__init__.py:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    entry.async_on_unload(entry.add_update_listener(async_update_options))
```
I couldn't see what this was for.

rianadon replied:
It looks like `async_setup_platforms` was something used in entities when I first wrote this code but is not replaced with `async_forward_entry_setups(entry, PLATFORMS)`.

The `add_update_listener` I believe is to trigger a refresh of entities if the options change. I still see many integrations using this line.

davet2001 commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,134 @@
+        errors = {}
+
+        if user_input is not None:
+            if user_input[CONF_HOST] in self._async_current_hosts():
+                return self.async_abort(reason="already_configured")
```
Could this be done using `_abort_if_unique_id_configured()` rather than a custom function?

davet2001 reviewed (commented):
Just an incomplete review to get started. Will continue later.

davet2001 reviewed (changes requested):
Thanks for the contribution @rianadon and also @SupremeSports for work on the earlier PR!

The bare minimum is test coverage for config_flow.py (which is done), but I suggest also adding tests for the actual sensor and DataUpdateCoordinator.  The tests should verify that data comes back in the places you expect it and can be done in a separate PR.

rianadon commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,134 @@
+                await self.async_set_unique_id(self.serial)
+            except httpx.HTTPError:
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
```
You're right. I don't think there should be any types other than HTTP Errors.

rianadon committed:
Address review comments

rianadon committed:
Add typings

rianadon committed:
Fix tests

rianadon commented:
Thank you for the review @davet2001. I think adding tests for the other components in a separate PR would be a good idea.

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted-new

rianadon commented:
@davet2001 @Quentame this is ready for another review.

davet2001 reviewed (changes requested):
Only minor comment. Looking good! I will request a second opinion in case I missed something,

rianadon committed:
Refactor: remove async_generate_schema

rianadon commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,120 @@
+        self.model = None
+
+    @callback
+    def _async_generate_schema(self):
```
Better yet, I took out this function method since it looks like very few other components have an `async_generate_schema` method. Instead, they all create the `vol.Schema` inside the `async_step_user`.

davet2001 commented:
@rianadon Yes, a function is probably only necessary if there's a need for it to change based on parameters.  Could use a constant if it's needed in more than one place.

I think a new constant deprecation has caused the test to fail, please could you take a look?

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted-new

rianadon committed:
Use UnitOfElectricPotentialVolt

rianadon commented:
Thank you! I've updated the constant.

davet2001 reviewed (approved):
Ok from me now. Thanks @rianadon! 👍
2nd opinion wanted (just in case I missed something).

rianadon commented:
@Quentame I see you started a review. Would you mind providing that 2nd review/opinion?

rianadon commented:
If there's anyone on the thread who would like to provide a second review, I'd really appreciate it.

It's been almost a month since the last review, and it would be great to have support for all TED devices on Home Assistant.

emontnemery commented on homeassistant/components/ted/sensor.py:
```diff
@@ -0,0 +1,196 @@
+
+def build_sensor_descs(name: str, prefix: str, stype: str, is_net: bool):
+    """Return a list of sensors with given key prefix and type (Production / Consumption)."""
+    if is_net:  # If the sensor represents a net energy
+        total_state_class = SensorStateClass.TOTAL
+    else:
+        total_state_class = SensorStateClass.TOTAL_INCREASING
```
I don't see how this works.

If the sensor is a net meter, and it's a daily or month to data type sensor, calculating a total requires `last_reset` to be set, but I don't see that in the code.
Please read this carefully: https://developers.home-assistant.io/docs/core/entity/sensor#entities-representing-a-total-amount

rianadon replied:
Thank you for taking a look at this @emontnemery. You are right... I was confused as to how the total state class worked. I took a read through the docs and hopefully the latest commit resolve this.

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted-new

rianadon committed:
Add last_reset to sensor, energy->power

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted-new

rianadon committed:
Merge remote-tracking branch 'upstream/dev' into ted-new

rianadon committed:
Delay expected date of retiring ted5000 integration

rianadon committed:
Sort manifest keys

rianadon commented:
Hi @emontnemery. As I mentioned before I had held off implementing the state class for net meters because I was a bit confused how to proceed. Your comment helped clarify things, but I'm still facing an underlying problem caused by the Ted 6000 device resetting at inconsistent times.

I had asked on the discord and got no response, so I figure I'll ask here.

I've been running my changes with `last_reset` for a few weeks, and today I decided to create a statistics card for one for one of the net meters to check how they're working.

<img width="503" alt="image" src="https://user-images.githubusercontent.com/18223213/221823094-69eba23e-64a3-459b-b1fe-837262460353.png">

This is what I got. Based on the form of the state measurement, the sum should be an ever-increasing line. However, it has jaggies because the time that the sensor resets is inconsistent. Sometimes it's at midnight, but other days it can be as far as five minutes behind.

As comparison, here's a non-net energy value that uses the `total_increasing` state class. Notice how the line is straight.

<img width="403" alt="image" src="https://user-images.githubusercontent.com/18223213/221829489-efa7b0ba-aa9b-4f49-8b88-699c4e4884dc.png">


Is there any great way to handle the inconsistent reset times? And if not, I'm thinking that maybe I should not give these net entities a state class `measurement`. The Ted6000 also publishes historical data (for example what was the total net energy use 6 days ago). Perhaps I could use these to detect whether the day has rolled over, but it sounds like a lot of extra logic. Perhaps this is better left to a future PR? It would be great to get this work merged.

emontnemery commented:
> it has jaggies because the time that the sensor resets is inconsistent. Sometimes it's at midnight, but other days it can be as far as five minutes behind.

The problem is that you're trying to calculate the `last_reset` time based on the local time. That simply won't work, you need to either get the time from the API or detect that a new period has started and bump the last_reset only then.
An example of the former is if the API includes some meta data indicating for which date the reading is.

Let's chat on discord if the way forward is not clear, I'm emontnemery#6618 there

rianadon commented:
A good reminder! I'm still working on this.

edenhaus commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+
+    VERSION = 1
+
+    def __init__(self):
+        """Initialize an TED flow."""
+        self.name = None
+        self.serial = None
+        self.model = None
```
Please use local variables instead as they are used only in one function

```suggestion
```

edenhaus commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+                    user_input[CONF_HOST], async_client=get_async_client(self.hass)
+                )
+                await reader.update()
+                self.serial = reader.gateway_id
+                self.model = 5000 if isinstance(reader, tedpy.TED5000) else 6000
+                await self.async_set_unique_id(self.serial)
+                self._abort_if_unique_id_configured({CONF_HOST: user_input[CONF_HOST]})
```
Please move it to the else block

edenhaus commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+            else:
+                data = user_input.copy()
+                data[CONF_NAME] = f"TED {self.model}"
+                return self.async_create_entry(title=data[CONF_NAME], data=data)
```
```suggestion
                return self.async_create_entry(title=data[CONF_NAME], data=data, options=OPTION_DEFAULTS)
```

edenhaus commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    self.create_option_schema_field(CONF_SPYDER_POWER_NOW): bool,
+                    self.create_option_schema_field(CONF_SPYDER_ENERGY_DAILY): bool,
+                    self.create_option_schema_field(CONF_SPYDER_ENERGY_MTD): bool,
+                    self.create_option_schema_field(CONF_MTU_POWER_VOLTAGE): bool,
+                    self.create_option_schema_field(CONF_MTU_POWER_NOW): bool,
+                    self.create_option_schema_field(CONF_MTU_ENERGY_DAILY): bool,
+                    self.create_option_schema_field(CONF_MTU_ENERGY_MTD): bool,
+                }
+            ),
```
```suggestion
            data_schema=self.add_suggested_values_to_schema(vol.Schema(
                {
                    vol.Required(CONF_SPYDER_POWER_NOW): bool,
                    vol.Required(CONF_SPYDER_ENERGY_DAILY): bool,
                    vol.Required(CONF_SPYDER_ENERGY_MTD): bool,
                    vol.Required(CONF_MTU_POWER_VOLTAGE): bool,
                    vol.Required(CONF_MTU_POWER_NOW): bool,
                    vol.Required(CONF_MTU_ENERGY_DAILY): bool,
                    vol.Required(CONF_MTU_ENERGY_MTD): bool,
                }
            ), self.config_entry.options),
```

edenhaus commented on homeassistant/components/ted/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+            ),
+        )
+
+    def create_option_schema_field(self, name: str):
+        """Create the schema for a specific TED option with the appropriate default value."""
+        return vol.Required(
+            name, default=self.config_entry.options.get(name, OPTION_DEFAULTS[name])
+        )
```
```suggestion
```

edenhaus commented on homeassistant/components/ted/coordinator.py:
```diff
@@ -0,0 +1,68 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class TedUpdateCoordinator(DataUpdateCoordinator):
```
```suggestion
class TedUpdateCoordinator(DataUpdateCoordinator[dict[str, Any]]):
```

edenhaus commented on homeassistant/components/ted/__init__.py:
```diff
@@ -0,0 +1,55 @@
+            async_client=get_async_client(hass),
+        )
+
+    coordinator = TedUpdateCoordinator(hass, ted_reader=ted_reader, name=f"TED {name}")
```
The coordinator would be `TED TED 600`. I think we can remove one `TED`

edenhaus commented on homeassistant/components/ted/sensor.py:
```diff
@@ -0,0 +1,218 @@
+    async_add_entities(entities)
+
+
+class TedSensor(CoordinatorEntity, SensorEntity):
```
For new integrations, we require the following:
- Support for [the new entity naming](https://developers.home-assistant.io/docs/core/entity#has_entity_name-true-mandatory-for-new-integrations)
- Services must be [translatable](https://developers.home-assistant.io/blog/2023/07/11/translating-services)
- Entities and their attributes should be [translatable](https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations/)

Please adopt your code to add support for it.

edenhaus commented on homeassistant/components/ted/coordinator.py:
```diff
@@ -0,0 +1,68 @@
+            except httpx.HTTPError as err:
+                raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+            data["type"] = self.ted_reader.system_type
```
Why are you creating a dict out of all data?
Can we not use a data class for it?

Ramias1 commented:
Hi @rianadon are you still working on this?  Looking forward to using this.  Thanks

rianadon commented:
Hey. I haven't had the time to finish this. I've been using it as a custom component for the time being, but it's a little buggy (see [this comment](https://github.com/home-assistant/core/pull/82959#issuecomment-1447947212)).

The big missing step is to query the device for the local time, and add some detection around whether it thinks it's a new day in order to correctly set `last_reset`. I just pushed my changes for reading time to the python library I wrote for the TED, but the CI on GitHub is failing, so I'm not going to publish in case that breaks something.

It'll be a while before I can get back to this, so if someone would like to take over the code I'd be happy to hand it off.
